File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math
def solution(fees, records):

    timeleft = {}
    
    for i in records: 
        time,carnum,state = i.split()
        if carnum not in timeleft.keys() :
            timeleft[carnum] = fees[0]#기본 주차 가능 시간 
            
    parkinglot = []
    timeremember = {}
    for i in records:
        time,carnum,state = i.split()
        realtime = int(time[0:2])*60 + int(time[3:])
        
        if state == "IN" :
            parkinglot.append(carnum)
            timeremember[carnum] = realtime
        
        else :
            #나갈때가 되었어요 . 
            parkinglot.remove(carnum)
            howmanytime = realtime-timeremember[carnum] #시간 차이
            timeleft[carnum] -= howmanytime
            
    # print(parkinglot) #아직 안빠진 애가 있다?
    
    if len(parkinglot) >0 :
        for left in parkinglot :
            beforetime = timeremember[left] 
            timeleft[left] -= (1439 - beforetime ) # 다 정산 완료 #여기가좀이상.
            
    print(timeleft)
    
    paydict = {}      
        
    for i in timeleft:
        if timeleft[i] >= 0 :
            paydict[i] = fees[1]
        elif timeleft[i]< 0 :
            cost = fees[1]+ math.ceil((-timeleft[i])/fees[2])*fees[3]
            paydict[i] = cost
            
    new = sorted(paydict.items(),key= lambda x:x[0])
    
    answer = []
    
    for i in new :
        answer.append(i[1])
   
    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;

class Solution {
    public Map<Integer, Integer> map = new HashMap<>();
    public Map<Integer, Integer> timemap = new HashMap<>();
    public int maxtime = 23*60+59;
    public int[] solution(int[] fees, String[] records) {
        
        for (int i = 0 ; i<records.length; i++)
        {
            String [] line = records[i].split(" ");
            String [] startarr = line[0].split(":");
            int time = Integer.parseInt(startarr[0])*60 + Integer.parseInt(startarr[1]);
            int carnum = Integer.parseInt(line[1]);
            if(line[2].equals("IN"))
            {
                //차번호, 입차시간.
                map.put(carnum,time);
            }
            else
            {
                int newtime = time-map.getOrDefault(carnum,0);
                timemap.put(carnum,newtime+timemap.getOrDefault(carnum,0));
                map.remove(carnum);
            }
        }
        
        //출차 안된애들 확인
        
        
        
        if(!map.isEmpty())
        {
            //계산
            int [] answer = new int [map.keySet().size()];
            List<Integer> notfinish = new LinkedList<>(map.keySet());
            
            for(Integer carnum : notfinish)
            {
                int newtime = maxtime-map.getOrDefault(carnum,0);
                timemap.put(carnum,newtime+timemap.getOrDefault(carnum,0));
                map.remove(carnum);
            }
        }
        
        //계산
        int [] answer = new int [timemap.keySet().size()];
        List<Integer> keylist = new LinkedList<>(timemap.keySet());
        
        Collections.sort(keylist);
        
        int idx = 0 ;
        for(Integer carnum : keylist)
            {
                int time = timemap.get(carnum);
                int val = 0 ;
                
                if (time<=fees[0])
                {
                    answer[idx] = fees[1];
                } 
                else
                {
                    answer[idx] = fees[1] + (int)Math.ceil((double)(time-fees[0])/fees[2])*fees[3];
                }
                idx +=1;
            }
        
        return answer;
    }
}